app/post.go

```diff
@@ -1174,3 +1174,131 @@ func (a *App) MaxPostSize() int {
 
 	return maxPostSize
 }
+
+func (a *App) countMentionsFromPost(user *model.User, post *model.Post) (int, *model.AppError) {
+	channel, err := a.GetChannel(post.ChannelId)
+	if err != nil {
+		return 0, err
+	}
+
+	if channel.Type == model.CHANNEL_DIRECT {
+		// Return -1 to indicate that the mention count should be equal to the post count
+		return -1, nil
+	}
+
+	channelMember, err := a.GetChannelMember(channel.Id, user.Id)
+	if err != nil {
+		return 0, err
+	}
+
+	keywords := addMentionKeywordsForUser(
+		map[string][]string{},
+		user,
+		channelMember.NotifyProps,
+		&model.Status{Status: model.STATUS_ONLINE}, // Assume the user is online since they would've triggered this
+		true, // Assume channel mentions are always allowed for simplicity
+	)
+	commentMentions := user.NotifyProps[model.COMMENTS_NOTIFY_PROP]
+	checkForCommentMentions := commentMentions == model.COMMENTS_NOTIFY_ROOT || commentMentions == model.COMMENTS_NOTIFY_ANY
+
+	// A mapping of thread root IDs to whether or not a post in that thread mentions the user
+	mentionedByThread := make(map[string]bool)
+
+	isCommentMention := func(post *model.Post, otherPosts map[string]*model.Post) bool {
+		if post.RootId == "" {
+			// Not a comment
+			return false
+		}
+
+		if mentioned, ok := mentionedByThread[post.RootId]; ok {
+			// We've already figured out if the user was mentioned by this thread
+			return mentioned
+		}
+
+		// Whether or not the user was mentioned because they started the thread
+		mentioned := otherPosts[post.RootId].UserId == user.Id
+
+		// Or because they commented on it before this post
```

Comment:
weren't this on the profilemap? maybe we should pass it and add them instead of going through it for every post in the thread?

Reply:
I meant that since we are going to do this on a bunch of users, shouldn't we take that group and for every message add the notification?

but now thinking about it, I realize again that there are two main paths (new message vs mark as unread) and each one of them has a different usecase with different problems, so I might be thinking how to improve one and that might hurt the other one (performance wise). With that in mind, we might want to separate those paths limiting code reuse, which also makes maintaining code a bit harder... but I feel like that's a problem beyond the scope of this PR, but not too far. we might want to work on it as soon as we land the feature. Let me know your thoughts or if we should discuss this elsewhere with more people

Reply:
I agree that that's a good idea to look at how these two paths work together. I tried to do a bit of that here with the changes to how we make the keyword map, but I didn't have a chance to do it for comment mentions because of how I was originally counting them (by iterating through the posts in order and keeping track of which of the user's comments we'd seen before). It's worth looking into more, but each code path (one user and multiple posts vs one post and multiple users) has different optimizations that can be done, so we probably can't get rid of all duplication here. It's something we can look into the future though